### PR TITLE
Consensus Algorithms: Move template functions into .h file.

### PR DESCRIPTION
**Pull request by bangerth**
When converting the consensus algorithms to take any kind of data (see #13722), one of the annoying changes was that all sorts of files suddenly had to `#include` the `.templates.h` file because the handful of existing explicit instantiations no longer covered all of the data types we now push through these functions. The right approach is to just move the implementations of these template functions into the `.h` file, and to remove the explicit instantiations. This patch does that.

In reference to #13208.

/rebuild

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]

**Comment by masterleinad** [comment]
There is a merge conflict with `source/base/mpi.cc`.

**Review by masterleinad — APPROVED** [review]

**Comment by bangerth** [comment]
Rebased.

**Review by kronbichler — APPROVED** [review]

File: include/deal.II/base/mpi_consensus_algorithms.templates.h [hunk 1]
```diff
@@ -1,925 +0,0 @@
...
-// The deal.II library is free software; you can use it, redistribute
-// it, and/or modify it under the terms of the GNU Lesser General
-// Public License as published by the Free Software Foundation; either
-// version 2.1 of the License, or (at your option) any later version.
-// The full text of the license can be found in the file LICENSE.md at
-// the top level directory of deal.II.
-//
-// ---------------------------------------------------------------------
-
-#ifndef dealii_mpi_consensus_algorithm_templates_h
```
Comment (peterrum) [inline include/deal.II/base/mpi_consensus_algorithms.templates.h:16]: Could we keep this file for now (empty) and instead add a warning?

**Review by peterrum — APPROVED** [review]

File: include/deal.II/base/mpi_consensus_algorithms.templates.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/base/mpi_consensus_algorithms.templates.h]: Yes, great suggestion. I should have thought of that before.

**Comment by bangerth** [comment]
I did as @peterrum suggested: Leave the `.templates.h` file in place, but deprecate it.

**Pull request merged**